### ISSUE TITLE
ProgressParams supports work done progress only

### ISSUE THREAD

**Issue by pisv** [eclipse/lsp4j#479]
Currently, `ProgressParams` supports work done progress only. It does not support other kinds of progress (e.g., partial result progress).

**Comment by pisv** [comment]
I can see a number of ways to fix it:

1. Change declaration of `ProgressParams.value` to

```java
	@JsonAdapter(JsonElementTypeAdapter.Factory)
	Object value
```

This is straightforward, but obviously lacks direct support for `WorkDoneProgressNotification`.

2. Change declaration of `ProgressParams.value` to

```java
	@JsonAdapter(ProgressNotificationAdapter.Factory)
	Object value
```

where `ProgressNotificationAdapter` would create either `WorkDoneProgressNotification` or `JsonElement`

3. Change declaration of `ProgressParams.value` to

```java
	@JsonAdapter(ProgressNotificationAdapter.Factory)
	Either<WorkDoneProgressNotification, Object> value
```

Any thoughts?

**Comment by jonahgraham** [comment]
@Dufgui can you have a look at this please as I understand it this is follow on to the work you did in #370

**Comment by jcompagner** [comment]
see my comment here:
https://github.com/eclipse/lsp4j/issues/480#issuecomment-786502109

The problem that i see is that i think the class WorkDoneProgressNotificationAdapter tries to map it bit more on actual types...
Which can be ofcourse nice that you are fully type safe, problem is that i think what can get over the line can be anything..

angular language service when in ivy mode send over these 2 request:

{"jsonrpc":"2.0","method":"$/progress","params":{"token":"ngcc","value":{"done":false,"configFilePath":"c:/Users/jcomp/git/servoy_master/servoy-eclipse/com.servoy.eclipse.ngclient.ui/node/tsconfig.json","message":"Running ngcc for project c:/Users/jcomp/git/servoy_master/servoy-eclipse/com.servoy.eclipse.ngclient.ui/node/tsconfig.json"}}

{"jsonrpc":"2.0","method":"$/progress","params":{"token":"ngcc","value":{"done":true,"configFilePath":"c:/Users/jcomp/git/servoy_master/servoy-eclipse/com.servoy.eclipse.ngclient.ui/node/tsconfig.json","success":true}}}}
{"jsonrpc":"2.0","method":"$/progress","params":{"token":"ngcc","value":{"done":true,"configFilePath":"c:/Users/jcomp/git/servoy_master/servoy-eclipse/com.servoy.eclipse.ngclient.ui/node/tsconfig.json","success":true}}}


the problem is that i think that can be anything. is "configFilePath" a protocol defined something? i don't think so
My feeling is that the "value" object can be anything

So i think that lsp4j must be a bit more "generic" and if "method":"$/progress" is encountered 
just parse the params and just ship the values as a plain Map<String,Object>

i "hacked" it a bit so i added this case in the switch statement:

case "done": {
					if (in.nextBoolean()) {
						kind = "end";
					}
					else {
						kind = "report";
					}
					break;
				}

so i map on the once that are then created.

But i think we should just have a ProgressNotification(token, map) class and that just json values that are send over
then the public void notifyProgress(ProgressParams params) just gets that class and the LanguageClientImpl implementation can handle that how ever it wants and it has all the data..

**Comment by szarnekow** [comment]
Please see also the comment here https://github.com/eclipse/lsp4j/pull/442#discussion_r483470415
which quotes the spec
> Work done progress is reported using the generic $/progress notification. The value payload of a work done progress notification can be of three different forms.

Is there an updated version of the spec available that indicates that the param can be of arbitrary shape?

**Comment by jcompagner** [comment]
i am just looking at this one:

https://microsoft.github.io/language-server-protocol/specifications/specification-current/#progress

and then

type ProgressToken = integer | string;

interface ProgressParams<T> {
	/**
	 * The progress token provided by the client or server.
	 */
	token: ProgressToken;

	/**
	 * The progress data.
	 */
	value: T;
}

thats what angular language server just sends over 

token is "ngcc"

with a value object that as far as i can see is just be anything?
Because what is T here?

**Comment by pisv** [comment]
> Work done progress is reported using the generic $/progress notification. The value payload of a work done progress notification can be of three different forms.

This quote talks about work done progress notifications. However, the specification states that

> The base protocol offers also support to report progress in a generic fashion. This mechanism can be used to report **any kind of progress** including work done progress (usually used to report progress in the user interface using a progress bar) and partial result progress to support streaming of results.

See also https://microsoft.github.io/language-server-protocol/specifications/specification-3-15/#partialResults.

In general, the progress data can be anything. And it has been so since LSP 3.15, when progress support was introduced.

**Comment by szarnekow** [comment]
Thank you for clarifying this. That part of the spec must have slipped through the cracks when the initial progress support landed in lsp4j.

**Comment by jonahgraham** [comment]
Hi folks, this issue was previously identified as being part of LSP4J 0.11.0 as it provided part of the support for completing LSP 3.16.0. If this is a non-blocking issue for 0.11.0 release, please update/remove milestone so that we can progress with the LSP4J 0.11.0 release soon. We can have a 0.12.0 release soon after the 0.11.0 release if there is value in that. In the absence of feedback I will remove the milestone in a few days so that we can release 0.11.0 as many consumers are waiting on an official LSP4J 0.11.0 release. Thank you.

**Comment by jcompagner** [comment]
We do need to have this fixed as soon as possible (but i guess it doesn't have to be in 0.11 if we can create a 0.12 soon after that)
because as an example if Angular Language Server makes IVY usage the default and people are using "strictTemplates:true" as a angular compiler option,  then WWD will generate a lot of exceptions in the log because of this problem.
Because then the angular server sends over a lot of those generic progress messages over the line.

**Comment by pisv** [comment]
Please note that a PR that fixes this issue was provided 7 days ago, but unfortunately it has not been reviewed yet. Can anyone please take a look? I'm willing to do more work if necessary to get it fixed in 0.11.0.

**Comment by jonahgraham** [comment]
In the absence of any committers with closer knowledge reviewing this code, I have done a review and this associate PR looks close to fine to me so I will push this through to completion with @pisv's help.
